### ISSUE TITLE
static _SCHEMA field usage instead of Interface property

### ISSUE THREAD

**Issue by asedovski** [confluentinc/confluent-kafka-dotnet#1330]
https://github.com/confluentinc/confluent-kafka-dotnet/blob/98f019aff4a622037a8f3ce0d769c59f47554c78/src/Confluent.SchemaRegistry.Serdes.Avro/SpecificSerializerImpl.cs#L70

Due to static field usage in this place it makes pointless having ISpecificRecord at all.

Could you please replace static field usage with ISpecificRecord.Schema property?

**Comment by mhowlett** [comment]
what problem is this causing?

I had a very quick look at this. to do as you suggest, we need an instance of T. to get this, I think we need to use reflection and assume a parameterless constructor - we can't use a `new()` constraint on `T` because `string` for instance doesn't have a parameterless constructor. having done this, we can access `ISpecificRecord.Schema`. Which in practice then always accesses the static `_SCHEMA` field in the avrogen made class. So we've effectively achieved nothing.

with that said, i agree this all seems a bit stupid.

happy to make changes to address any real issue, but please provide more info on that.

**Comment by asedovski** [comment]
Problem is it is not enough to implement ISpecificRecord to be able serialize object because Schema from ISpecificRecord is not used. So my class SHOULD implement ISpecificRecord and SHOULD have static field. In my case I have base class who is responsible for building valid schema but all derived class still should have static field to be able serialized what is totally messy :)

In real method where serialization is happening we have T data, maybe it make sens to shift initialization of writerSchemaString at this place but do it only once at first serialization.

For now if I have valid implementation of ISpecificRecord but dont have static _SCHEMA I am receiving NullReferenceException when try to serialize message...

**Comment by BillyFromAHill** [comment]
I agree with @asedovski. It's a very strange behaviour. Even at runtime, library user doesn't know what's wrong with a serializing model if it doesn't have public static field named _SCHEMA. I understand that current implementation structure needs something like "static fields for interfaces" for fast fix. I suggest something like this for a moment https://github.com/confluentinc/confluent-kafka-dotnet/pull/1377

**Comment by BillyFromAHill** [comment]
I'll try to do something more appropriate next week.

**Comment by mhowlett** [comment]
Thanks @BillyFromAHill - good idea. And thanks @asedovski for articulating. Would happily accept a PR along the lines of what I was thinking.

**Comment by BillyFromAHill** [comment]
Made some refactoring here https://github.com/confluentinc/confluent-kafka-dotnet/pull/1383 Added ISpecificRecord instance creation like @mhowlett suggested and like it implemented inside Avro.Specific.ObjectCreator. Added exception in case type has no parameterless constructor. It's correct behavior because if there is no such constructor, exception will be raised from Avro.Specific.ObjectCreator, so we do it earlier and with a nice description.

**Comment by KalleOlaviNiemitalo** [comment]
I'd expect a type without a parameterless constructor to work for serializing, if the app doesn't need to deserialize it. That seems a niche scenario though, as types generated by `avrogen` do have parameterless constructors.
